georiviere/main/signals.py

```diff
@@ -0,0 +1,59 @@
+from django.contrib.gis.db.models.functions import Distance
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import F, FloatField, Case, When
+
+from georiviere.functions import ClosestPoint, Length, LineSubString, LineLocatePoint
+from georiviere.river.models import Stream
+from georiviere.main.models import DistanceToSource
+
+
+def annotate_distance_to_source(streams, instance):
+    if streams:
+        streams = streams.annotate(
+            locate_source=LineLocatePoint(F('geom'), F('source_location')),
+            locate_object=LineLocatePoint(F('geom'), ClosestPoint(F('geom'), instance.geom)),
+            locate=Length(LineSubString(F('geom'),
+                                        Case(
+                                            When(locate_source__gte=F('locate_object'), then=F('locate_object')),
+                                            default=F('locate_source')),
+                                        Case(
+                                            When(locate_source__gte=F('locate_object'), then=F('locate_source')),
+                                            default=F('locate_object')))) + Distance(
+                instance.geom,
+                F('geom'),
+                output_field=FloatField()) + Distance(
+                F('geom'),
+                F('source_location'),
+                output_field=FloatField())
+        )
+    return streams
+
+
+def save_objects_generate_distance_to_source(sender, instance, **kwargs):
+    if not hasattr(instance, 'get_topology'):
+        streams = annotate_distance_to_source(instance.streams, instance)
+        for stream in streams:
+            DistanceToSource.objects.update_or_create(
+                object_id=instance.pk,
+                content_type=ContentType.objects.get_for_model(instance._meta.model),
```

Comment:
I think you can instantiate with content_object=instance instead of object_id and content_type